hw/ip/otbn/dv/otbnsim/sim/stats.py

```diff
@@ -0,0 +1,302 @@
+
+    def record_stall(self) -> None:
+        self.stall_count += 1
+
+    def record_insn(self, pc: int, insn: OTBNInsn,
+                    state_before_commit: OTBNState) -> None:
+        '''Collect execution statistics'''
+
+        self.insn_count += 1
+
+        # Instruction histogram
+        self.instruction_histo[insn.insn.mnemonic] += 1
+
+        # Function calls
+        # - Direct function calls: jal x1, <offset>
+        # - Indirect function calls: jalr x1, <grs1>, 0
+        if isinstance(insn, JAL) and insn.grd == 1:
+            call_stack = state_before_commit.peek_call_stack()
+            if call_stack:
+                caller_func = call_stack[0]
+            else:
+                caller_func = state_before_commit.start_address
+
+            self.func_calls.append({
+                'call_site': pc,
+                'caller_func': caller_func,
+                'callee_func': state_before_commit.pc_next,
+            })
+
+        # Loops
+        if isinstance(insn, LOOPI):
+            self.loops.append({
+                'loop_addr': pc,
+                'loop_len': insn.bodysize,
+                'iterations': insn.iterations,
+            })
+
+        if isinstance(insn, LOOP):
+            iters = state_before_commit.gprs.get_reg(insn.grs).read_unsigned()
+            self.loops.append({
+                'loop_addr': pc,
+                'loop_len': insn.bodysize,
+                'iterations': iters,
+            })
+
+
+def _dwarf_decode_file_line(dwarf_info: DWARFInfo,
+                            address: int) -> Optional[Tuple[str, int]]:
+    # Go over all the line programs in the DWARF information, looking for
+    # one that describes the given address.
+    for CU in dwarf_info.iter_CUs():
+        # First, look at line programs to find the file/line for the address
+        lineprog = dwarf_info.line_program_for_CU(CU)
+        prevstate = None
+        for entry in lineprog.get_entries():
+            # We're interested in those entries where a new state is assigned
+            if entry.state is None:
+                continue
+            if entry.state.end_sequence:
+                # if the line number sequence ends, clear prevstate.
+                prevstate = None
+                continue
+            # Looking for a range of addresses in two consecutive states that
+            # contain the required address.
+            if prevstate and prevstate.address <= address < entry.state.address:
+                filename = lineprog['file_entry'][prevstate.file - 1].name.decode('utf-8')
+                line = prevstate.line
+                return filename, line
+            prevstate = entry.state
+    return None
+
+
+def _get_addr_symbol_map(elf_file: ELFFile) -> Dict[int, str]:
+    section = elf_file.get_section_by_name('.symtab')
+
+    if not section:
+        return {}
+
+    if not isinstance(section, SymbolTableSection):
+        return {}
```

Comment:
Nit: Since `None` isn't a `SymbolTableSection`, I don't think we actually need the first test. Maybe this would be better with just a comment?